// @vtaits/react-color-picker/lib/HueSpectrum.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _common = _interopRequireWildcard(require("./utils/common"));

var _defaultColor = _interopRequireDefault(require("./defaultColor"));

var _validate = _interopRequireDefault(require("./utils/validate"));

var _toColorValue = _interopRequireDefault(require("./utils/toColorValue"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var HueSpectrum = /*#__PURE__*/function (_BaseComponent) {
  (0, _inherits2.default)(HueSpectrum, _BaseComponent);

  var _super = _createSuper(HueSpectrum);

  function HueSpectrum() {
    var _this;

    (0, _classCallCheck2.default)(this, HueSpectrum);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", _objectSpread(_objectSpread({}, _common.baseInitialState), {}, {
      h: 0
    }));
    return _this;
  }

  (0, _createClass2.default)(HueSpectrum, [{
    key: "isComponentMounted",
    value: function isComponentMounted() {
      return this.mounted;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.mounted = true;
      this.updateDragPositionIf();
    }
  }, {
    key: "updateDragPositionIf",
    value: function updateDragPositionIf() {
      if (!this.props.height) {
        this.setState({});
      }
    }
  }, {
    key: "getDragPosition",
    value: function getDragPosition() {
      var _this$props = this.props,
          height = _this$props.height,
          pointerSize = _this$props.pointerSize;

      if (!height && !this.isComponentMounted()) {
        return null;
      }

      var computedHeight = height || this.getDOMRegion().getHeight();
      var size = pointerSize;
      var pos = Math.round(this.hsv.h * computedHeight / 360);
      var diff = Math.round(size / 2);
      return pos - diff;
    }
  }, {
    key: "updateColor",
    value: function updateColor(point) {
      var newPoint = (0, _validate.default)(point);
      this.hsv.h = newPoint.y * 360 / newPoint.height;

      var newHsv = _objectSpread({}, this.hsv);

      var newH;

      if (this.hsv.h !== 0) {
        newH = this.hsv.h;
      }

      newH = this.hsv.h !== 0 ? this.hsv.h : 0;
      this.setState({
        h: newH
      });
      this.hsv = newHsv;
      return newHsv;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          style = _this$props2.style,
          value = _this$props2.value,
          defaultColor = _this$props2.defaultColor,
          pointerSize = _this$props2.pointerSize,
          height = _this$props2.height,
          width = _this$props2.width;
      var h = this.state.h;
      this.hsv = (0, _toColorValue.default)(value || defaultColor);

      if (h === 360 && !this.hsv.h) {
        // in order to show bottom red as well on drag
        this.hsv.h = 360;
      }

      var rootStyle = _objectSpread({}, style);

      if (height) {
        rootStyle.height = height;
      }

      if (width) {
        rootStyle.width = width;
      }

      var dragStyle = {
        height: pointerSize
      };
      var dragPos = this.getDragPosition();

      if (dragPos !== null) {
        dragStyle.top = dragPos;
        dragStyle.display = 'block';
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        className: "react-color-picker__hue-spectrum",
        style: rootStyle,
        onMouseDown: this.onMouseDown,
        role: "button",
        ref: this.rootRef,
        tabIndex: 0
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "react-color-picker__hue-drag",
        style: dragStyle
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "react-color-picker__hue-inner"
      })));
    }
  }]);
  return HueSpectrum;
}(_common.default);

(0, _defineProperty2.default)(HueSpectrum, "propTypes", _objectSpread(_objectSpread({}, _common.basePropTypes), {}, {
  value: _propTypes.default.any,
  height: _propTypes.default.number,
  width: _propTypes.default.number,
  pointerSize: _propTypes.default.number,
  defaultColor: _propTypes.default.any,
  isHueSpectrum: _propTypes.default.bool
}));
(0, _defineProperty2.default)(HueSpectrum, "defaultProps", _objectSpread(_objectSpread({}, _common.baseDefaultProps), {}, {
  value: null,
  height: 300,
  width: 30,
  pointerSize: 3,
  defaultColor: _defaultColor.default,
  isHueSpectrum: true
}));
var _default = HueSpectrum;
exports.default = _default;
