// @vtaits/react-color-picker/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "HueSpectrum", {
  enumerable: true,
  get: function get() {
    return _HueSpectrum.default;
  }
});
Object.defineProperty(exports, "SaturationSpectrum", {
  enumerable: true,
  get: function get() {
    return _SaturationSpectrum.default;
  }
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _HueSpectrum = _interopRequireDefault(require("./HueSpectrum"));

var _SaturationSpectrum = _interopRequireDefault(require("./SaturationSpectrum"));

var _toColorValue = _interopRequireDefault(require("./utils/toColorValue"));

var _toStringValue = _interopRequireDefault(require("./utils/toStringValue"));

var _defaultColor = _interopRequireDefault(require("./defaultColor"));

var _excluded = ["color", "className", "hueStyle", "hueHeight", "hueMargin", "hueWidth", "defaultColor", "value", "saturationHeight", "saturationWidth"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ColorPicker = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ColorPicker, _Component);

  var _super = _createSuper(ColorPicker);

  function ColorPicker() {
    var _this;

    (0, _classCallCheck2.default)(this, ColorPicker);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      dragHue: null
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSaturationChange", function (color) {
      _this.handleChange(color);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleHueChange", function (color) {
      _this.handleChange(color);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleHueDrag", function (hsv) {
      _this.setState({
        dragHue: hsv.h
      });

      _this.handleDrag(hsv);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSaturationDrag", function (hsv) {
      _this.handleDrag(hsv);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleHueMouseDown", function (hsv) {
      _this.setState({
        dragHue: hsv.h
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSaturationMouseDown", function (hsv) {
      _this.setState({
        dragHue: hsv.h
      });
    });
    return _this;
  }

  (0, _createClass2.default)(ColorPicker, [{
    key: "handleDrag",
    value: function handleDrag(color) {
      var onDrag = this.props.onDrag;
      onDrag((0, _toStringValue.default)(color), color);
    }
  }, {
    key: "handleChange",
    value: function handleChange(color) {
      var onChange = this.props.onChange;
      this.setState({
        dragHue: null
      });

      var newColor = _objectSpread({}, color);

      var value = (0, _toStringValue.default)(newColor);
      onChange(value, newColor);
    }
  }, {
    key: "render",
    value: function render() {
      var props = this.props;
      var color = props.color,
          propsClassName = props.className,
          propsHueStyle = props.hueStyle,
          hueHeight = props.hueHeight,
          hueMargin = props.hueMargin,
          hueWidth = props.hueWidth,
          defaultColor = props.defaultColor,
          propsValue = props.value,
          saturationHeight = props.saturationHeight,
          saturationWidth = props.saturationWidth,
          divProps = (0, _objectWithoutProperties2.default)(props, _excluded);
      var dragHue = this.state.dragHue;
      var className = (0, _classnames.default)(propsClassName, 'cp react-color-picker');

      var hueStyle = _objectSpread(_objectSpread({}, propsHueStyle), {}, {
        marginLeft: hueMargin
      });

      var value = (0, _toColorValue.default)(propsValue || defaultColor);
      var children = props.children;
      var hueSpectrumProps = null;
      var saturationSpectrumProps = null;

      if (children) {
        children = _react.default.Children.toArray(children).forEach(function (child) {
          if (child && child.props) {
            if (child.props.isHueSpectrum) {
              hueSpectrumProps = _objectSpread({}, child.props);
            }

            if (child.props.isSaturationSpectrum) {
              saturationSpectrumProps = _objectSpread({}, child.props);
            }
          }
        });
      }

      var saturationConfig = _objectSpread({
        onDrag: this.handleSaturationDrag,
        onChange: this.handleSaturationChange,
        onMouseDown: this.handleSaturationMouseDown
      }, saturationSpectrumProps);

      if (saturationConfig.width === undefined) {
        saturationConfig.width = saturationWidth;
      }

      if (saturationConfig.height === undefined) {
        saturationConfig.height = saturationHeight;
      }

      saturationConfig.inPicker = true;

      var hueConfig = _objectSpread({
        onDrag: this.handleHueDrag,
        onChange: this.handleHueChange,
        onMouseDown: this.handleHueMouseDown,
        style: hueStyle
      }, hueSpectrumProps);

      if (hueConfig.width === undefined) {
        hueConfig.width = hueWidth;
      }

      if (hueConfig.height === undefined) {
        hueConfig.height = hueHeight || saturationHeight;
      }

      hueConfig.inPicker = true;

      if (dragHue) {
        value.h = dragHue;
      }

      saturationConfig.value = _objectSpread({}, value);
      hueConfig.value = _objectSpread({}, value);
      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, divProps, {
        className: className
      }), /*#__PURE__*/_react.default.createElement(_SaturationSpectrum.default, saturationConfig), /*#__PURE__*/_react.default.createElement(_HueSpectrum.default, hueConfig));
    }
  }]);
  return ColorPicker;
}(_react.Component);

(0, _defineProperty2.default)(ColorPicker, "propTypes", {
  onDrag: _propTypes.default.func,
  onChange: _propTypes.default.func,
  value: _propTypes.default.any,
  defaultColor: _propTypes.default.any,
  hueHeight: _propTypes.default.number,
  hueMargin: _propTypes.default.number,
  hueWidth: _propTypes.default.number,
  saturationWidth: _propTypes.default.number,
  saturationHeight: _propTypes.default.number
});
(0, _defineProperty2.default)(ColorPicker, "defaultProps", {
  onDrag: Function.prototype,
  onChange: Function.prototype,
  value: null,
  defaultColor: _defaultColor.default,
  hueHeight: null,
  hueMargin: 10,
  hueWidth: 30,
  saturationWidth: 300,
  saturationHeight: 300
});
var _default = ColorPicker;
exports.default = _default;
