// @vtaits/react-color-picker/lib/utils/color.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
exports.fromRatio = fromRatio;
exports.toAlpha = toAlpha;
exports.toColor = toColor;
exports.toHsv = toHsv;
exports.toPure = toPure;

var _tinycolor = _interopRequireDefault(require("tinycolor2"));

function toColor(color) {
  return (0, _tinycolor.default)(color);
}

function toPure(color) {
  var _toColor$toHsl = toColor(color).toHsl(),
      h = _toColor$toHsl.h;

  return toColor({
    h: h,
    s: 100,
    l: 50,
    a: 1
  });
}

function fromRatio(color) {
  return _tinycolor.default.fromRatio(color);
}

function toAlpha(color, alpha) {
  var newAlpha = alpha > 1 ? alpha / 100 : alpha;
  var newColor = toColor(color).toRgb();
  newColor.a = newAlpha;
  return toColor(newColor);
}

function toHsv(color) {
  return toColor(color).toHsv();
}

var _default = {
  fromRatio: fromRatio,
  toAlpha: toAlpha,
  toColor: toColor,
  toHsv: toHsv,
  toPure: toPure
};
exports.default = _default;
