// @vtaits/react-color-picker/lib/utils/common.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.basePropTypes = exports.baseInitialState = exports.baseDefaultProps = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _region = _interopRequireDefault(require("region"));

var _dragHelper = _interopRequireDefault(require("drag-helper"));

var _toStringValue = _interopRequireDefault(require("./toStringValue"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var baseInitialState = {
  top: 0,
  left: 0,
  mouseDown: null
};
exports.baseInitialState = baseInitialState;
var basePropTypes = {
  inPicker: _propTypes.default.bool,
  onMouseDown: _propTypes.default.func,
  onDrag: _propTypes.default.func,
  onChange: _propTypes.default.func
};
exports.basePropTypes = basePropTypes;
var baseDefaultProps = {
  inPicker: false,
  onMouseDown: null,
  onDrag: null,
  onChange: null
};
exports.baseDefaultProps = baseDefaultProps;

var getEventInfo = function getEventInfo(event, region) {
  var x = event.clientX - region.left;
  var y = event.clientY - region.top;
  return {
    x: x,
    y: y,
    width: region.getWidth(),
    height: region.getHeight()
  };
};

var BaseComponent = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(BaseComponent, _Component);

  var _super = _createSuper(BaseComponent);

  function BaseComponent() {
    var _this;

    (0, _classCallCheck2.default)(this, BaseComponent);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", baseInitialState);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "rootRef", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDragStart", Function.prototype);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onMouseDown", function (event) {
      event.preventDefault();

      var region = _this.getDOMRegion();

      var info = getEventInfo(event, region);
      (0, _dragHelper.default)(event, {
        scope: (0, _assertThisInitialized2.default)(_this),
        constrainTo: region,

        /* eslint-disable no-param-reassign */
        onDragStart: function onDragStart(dragStartEvent, config) {
          config.initialPoint = info;
          config.minLeft = 0;
          config.maxLeft = region.width;
          this.handleDragStart(dragStartEvent);
        },

        /* eslint-enable no-param-reassign */
        onDrag: function onDrag(dragEvent, config) {
          var dragInfo = getEventInfo(dragEvent, region);
          var newHsv = this.updateColor(dragInfo);
          this.handleDrag(dragEvent, config, newHsv);
        },
        onDrop: function onDrop(dropEvent, config) {
          var dropInfo = getEventInfo(dropEvent, region);
          var newHsv = this.updateColor(dropInfo);
          this.handleDrop(dropEvent, config, newHsv);
        }
      });

      _this.updateColor(info);

      _this.handleMouseDown(event, {
        initialPoint: info
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMouseDown", function (event, config) {
      var onMouseDown = _this.props.onMouseDown;

      if (onMouseDown) {
        onMouseDown.apply((0, _assertThisInitialized2.default)(_this), _this.getColors(_this.hsv));
      }

      _this.handleDrag(event, config, _this.hsv);
    });
    return _this;
  }

  (0, _createClass2.default)(BaseComponent, [{
    key: "getDOMRegion",
    value: function getDOMRegion() {
      return _region.default.fromDOM(this.rootRef.current);
    }
  }, {
    key: "getColors",
    value: function getColors(hsv) {
      var inPicker = this.props.inPicker;
      var first = inPicker ? hsv : (0, _toStringValue.default)(hsv);
      var args = [first];

      if (!inPicker) {
        args.push(_objectSpread({}, hsv));
      }

      return args;
    }
  }, {
    key: "handleUpdate",
    value: function handleUpdate(event, config) {
      var diff = config.diff || {
        top: 0,
        left: 0
      };
      var initialPoint = config.initialPoint;

      if (initialPoint) {
        var left;
        left = initialPoint.x + diff.left;
        var top = initialPoint.y + diff.top;
        left = Math.max(left, config.minLeft);
        left = Math.min(left, config.maxLeft);
        this.setState({
          top: top,
          left: left,
          mouseDown: {
            x: left,
            y: top,
            width: initialPoint.width,
            height: initialPoint.height
          }
        });
      }
    }
  }, {
    key: "handleDrag",
    value: function handleDrag(event, config, hsv) {
      var onDrag = this.props.onDrag;
      this.handleUpdate(event, config);

      if (onDrag) {
        onDrag.apply(this, this.getColors(hsv));
      }
    }
  }, {
    key: "handleDrop",
    value: function handleDrop(event, config, hsv) {
      var onChange = this.props.onChange;
      this.handleUpdate(event, config);
      this.setState({
        mouseDown: false
      });

      if (onChange) {
        onChange.apply(this, this.getColors(hsv));
      }
    }
  }]);
  return BaseComponent;
}(_react.Component);

(0, _defineProperty2.default)(BaseComponent, "propTypes", basePropTypes);
(0, _defineProperty2.default)(BaseComponent, "defaultProps", baseDefaultProps);
var _default = BaseComponent;
exports.default = _default;
